pkg/scale/decode.go

```diff
@@ -478,51 +478,79 @@ func (ds *decodeState) decodeBool(dstv reflect.Value) (err error) {
-	b, err := ds.ReadByte()
+	const maxUint32 = ^uint32(0)
+	const maxUint64 = ^uint64(0)
+	prefix, err := ds.ReadByte()
 	if err != nil {
-		return
+		return fmt.Errorf("reading from buffer: %w", err)
 	}
 
 	in := dstv.Interface()
 	temp := reflect.New(reflect.TypeOf(in))
 	// check mode of encoding, stored at 2 least significant bits
-	mode := b & 3
-	switch {
-	case mode <= 2:
-		var val int64
-		val, err = ds.decodeSmallInt(b, mode)
+	mode := prefix % 4
+	var value uint64
+	switch mode {
+	case 0:
+		value = uint64(prefix >> 2)
+	case 1:
+		buf, err := ds.ReadByte()
 		if err != nil {
-			return
+			return fmt.Errorf("reading from buffer: %w", err)
 		}
-		temp.Elem().Set(reflect.ValueOf(val).Convert(reflect.TypeOf(in)))
-		dstv.Set(temp.Elem())
-	default:
-		// >4 byte mode
-		topSixBits := b >> 2
-		byteLen := uint(topSixBits) + 4
-
+		value = uint64(binary.LittleEndian.Uint16([]byte{prefix, buf}) >> 2)
+		if value <= 0b0011_1111 || value > 0b0111_1111_1111_1111 {
+			return fmt.Errorf("%w: %d (%b)", ErrU16OutOfRange, value, value)
+		}
+	case 2:
+		buf := make([]byte, 3)
+		_, err = ds.Read(buf)
+		if err != nil {
+			return fmt.Errorf("reading from buffer: %w", err)
+		}
+		value = uint64(binary.LittleEndian.Uint32(append([]byte{prefix}, buf...)) >> 2)
+		if value <= 0b0011_1111_1111_1111 || value > uint64(maxUint32>>2) {
+			return fmt.Errorf("%w: %d (%b)", ErrU32OutOfRange, value, value)
+		}
+	case 3:
+		byteLen := (prefix >> 2) + 4
 		buf := make([]byte, byteLen)
 		_, err = ds.Read(buf)
 		if err != nil {
-			return
+			return fmt.Errorf("reading from buffer: %w", err)
 		}
-
-		var o uint64
-		if byteLen == 4 {
-			o = uint64(binary.LittleEndian.Uint32(buf))
-		} else if byteLen > 4 && byteLen <= 8 {
+		switch byteLen {
+		case 4:
+			value = uint64(binary.LittleEndian.Uint32(buf))
+			if value <= uint64(maxUint32>>2) {
+				return ErrU32OutOfRange
+			}
+		case 8:
+			uintSize := 32 << (^uint(0) >> 32 & 1)
+			if uintSize == 32 {
+				return ErrU64NotSupported
+			}
 			tmp := make([]byte, 8)
 			copy(tmp, buf)
-			o = binary.LittleEndian.Uint64(tmp)
-		} else {
-			err = errors.New("could not decode invalid integer")
-			return
+			value = binary.LittleEndian.Uint64(tmp)
+			if value <= maxUint64>>8 {
+				return ErrU64OutOfRange
+			}
+		default:
+			return fmt.Errorf("unexpected prefix decoding compact uint: %d", prefix)
+
 		}
-		dstv.Set(reflect.ValueOf(o).Convert(reflect.TypeOf(in)))
 	}
+	temp.Elem().Set(reflect.ValueOf(value).Convert(reflect.TypeOf(in)))
+	dstv.Set(temp.Elem())
 	return
 }
 
+var ErrU16OutOfRange = errors.New("uint16 out of range")
+var ErrU32OutOfRange = errors.New("uint32 out of range")
+var ErrU64OutOfRange = errors.New("uint64 out of range")
+var ErrU64NotSupported = errors.New("uint64 is not supported")
+
 // decodeLength is helper method which calls decodeUint and casts to int
 func (ds *decodeState) decodeLength() (l int, err error) {
```

Comment:
```suggestion
func (ds *decodeState) decodeLength() (l uint, err error) {
```
can you change this to `uint`